### PR TITLE
Weekly modulations

### PR DESCRIPTION
**Pull request by semohr**
Created a pull request to do a code review.

### CODE REVIEW

File: covid19_npis/model/number_of_tests.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 import logging
 import pymc4 as pm
 import numpy as np
+import pandas as pd
```
Comment (semohr) [inline covid19_npis/model/number_of_tests.py]: We should not need pandas in this submodule. Convertion to tensors should be done in the modelParams.

File: covid19_npis/model/number_of_tests.py [hunk 2]
```diff
@@ -15,10 +16,85 @@
...
+    week_modulation_type : str
+        The type of modulation, accepts ``"step"`` or  ``"abs_sine`` (the default).
+    pr_mean_weekend_factor : float
+        Sets the prior mean of the factor :math:`f_w` by which weekends are counted.
+    pr_sigma_weekend_factor : float
+        Sets the prior sigma of the factor :math:`f_w` by which weekends are counted.
+    weekend_days : tuple of ints
+        The days counted as weekend if ``week_modulation_type`` is ``"step"``
+    model : :class:`Cov19Model`
+        if none, it is retrieved from the context
```
Comment (semohr) [inline covid19_npis/model/number_of_tests.py]: Docstring needs some changes, i.e. the parameters are not right anymore. Same for return values.

File: covid19_npis/model/number_of_tests.py [hunk 3]
```diff
@@ -15,10 +16,85 @@
...
+        loc=0,
+        scale=2,
+        conditionally_independent=True,
+        event_stack=(1, modelParams.num_countries, 1),
+        shape_label=(None, "country", None),
+        # transform=transformations.SoftPlus(scale=0.5),
+    )
+    weight = tf.math.sigmoid(weight_cross)
+
+    t = pd.date_range(start=modelParams.data_begin,end=modelParams.data_end).weekday    # create array with weekdays
```
Comment (semohr) [inline covid19_npis/model/number_of_tests.py]: Getting weekdays from modelParams should be the better approach. There should be a property for that, if not we need to create one.

File: covid19_npis/model/number_of_tests.py [hunk 4]
```diff
@@ -15,10 +16,85 @@
...
+    f = (1 - weight) * (1 - tf.math.abs(tf.math.sin(t[np.newaxis,...,np.newaxis,np.newaxis] / 7 * np.pi + offset / 2))) # modulation factor
```
Comment (semohr) [inline covid19_npis/model/number_of_tests.py]: tf.newaxis != np.newaxis not sure if it makes a difference tho.
Same for np.pi, there should be tf.pi.

We have to move the datearray to tensorflow for that, this is important **if** we want to run the model on a gpu at some point!

File: covid19_npis/model/number_of_tests.py [hunk 5]
```diff
@@ -15,10 +16,85 @@
...
+    new_cases_inferred = cases * (1 - f)    # total modulation
+
```
Comment (semohr) [inline covid19_npis/model/number_of_tests.py:92]: Add the new_cases_inferred to the trace via pm.deterministic and name from functin parameters.

File: covid19_npis/model/number_of_tests.py [hunk 6]
```diff
@@ -117,8 +193,18 @@ def generate_testing(name_total, name_positive, modelParams, new_E_t):
...
+        name="weekly_modulation",
+        modelParams=modelParams,
+        cases=positive_tests,
+        week_modulation_type="abs_sine",
+    )
+
     positive_tests = yield Deterministic(
-        name_positive, positive_tests, shape_label=("time", "country", "age_group")
+        name=name_positive,
+        value=positive_tests,
```
Comment (semohr) [inline covid19_npis/model/number_of_tests.py]: Move  yield Deterministic into weekly_modulation to unclutter higher level functions.

File: covid19_npis/model/number_of_tests.py [hunk 7]
```diff
@@ -15,10 +16,85 @@
...
+    -------
+    new_cases : :class:`~theano.tensor.TensorVariable`
+    """
+
+    log.debug("Week modulation")
+
+    """
+    # TODO: - general check of function
+            - which shape does the modulation have to be? i guess time,country,age?
+            - check prior parameters
```
Comment (semohr) [inline covid19_npis/model/number_of_tests.py:54]: Should be fine, we can revise that later on.

File: covid19_npis/model/number_of_tests.py [hunk 8]
```diff
@@ -15,10 +16,85 @@
...
+            - check: which ones should be modulated? (mostly: positive tests!)
```
Comment (semohr) [inline covid19_npis/model/number_of_tests.py]: For now positive tests, but we can in theory input every parameter with the same dimensions. ("time", "country", "age_group")

File: covid19_npis/model/number_of_tests.py [hunk 9]
```diff
@@ -15,10 +16,85 @@
...
+            - different modulations across: age, country?
```
Comment (semohr) [inline covid19_npis/model/number_of_tests.py:55]: That should work. Not sure if we need age groups, we have to try that.

File: covid19_npis/model/number_of_tests.py [hunk 10]
```diff
@@ -15,10 +16,85 @@
...
+            - check: are (cumulative) case numbers same as in unmodulated case? need some kind of normalization?
+            - store and plot parameters at end
```
Comment (semohr) [inline covid19_npis/model/number_of_tests.py:57]: You should be able to add the name of the parameter to the plotting list (dummy_model.py bottom).

**Review by semohr — COMMENTED** [review]
Added some comments

File: covid19_npis/model/number_of_tests.py (hunk 1 above)
Comment (woxxel) [reply covid19_npis/model/number_of_tests.py]: resolved - I didn't find any function/variable in the dataframe that already covered this and added an according function "get_weekdays" to the modelParams class

File: covid19_npis/model/number_of_tests.py (hunk 3 above)
Comment (woxxel) [reply covid19_npis/model/number_of_tests.py]: see above response - added "get_weekdays" to modelParams-class

File: covid19_npis/model/number_of_tests.py (hunk 4 above)
Comment (woxxel) [reply covid19_npis/model/number_of_tests.py]: changed accordingly. tf.pi does not exist, tf.constant(np.pi) seems to be the way to go.
i'm not entirely happy with the tf.newaxis part - there should be a more elegant way to broadcast it, but i couldnt get it done with tfs broadcast_to-functionality

File: covid19_npis/model/number_of_tests.py (hunk 5 above)
Comment (woxxel) [reply covid19_npis/model/number_of_tests.py:92]: done

File: covid19_npis/model/number_of_tests.py (hunk 6 above)
Comment (woxxel) [reply covid19_npis/model/number_of_tests.py]: done - though ... why does it need a "yield" then... this thing is really crunching my gears...

File: covid19_npis/model/number_of_tests.py (hunk 8 above)
Comment (woxxel) [reply covid19_npis/model/number_of_tests.py]: it's general enough such that you can call it on every test statistics

File: covid19_npis/model/number_of_tests.py (hunk 9 above)
Comment (woxxel) [reply covid19_npis/model/number_of_tests.py:55]: would only need an additional dimension to the weight and offset-parameter - shouldnt be much of a hassle

**Pull request merged**